File: BlinkyBle/BlinkyBle/ViewController.m
//
//  ViewController.m
//  BlinkyBle
//
//  Created by Hoan on 2014-08-18.
//  Copyright (c) 2014 I-SYST inc. All rights reserved.
//

#import "ViewController.h"

#define BLINKY_SERVICE_UUID             @"00000001-59a4-4d2b-8479-8bbdbcf77fcd"
#define BLINKY_CHAR_UUID                @"00000002-59a4-4d2b-8479-8bbdbcf77fcd"
#define DEVINFO_SERIVE_UUID             @"180A"

typedef struct __attribute__((packed)) {
    uint8_t pin;
    uint8_t cmd;
} IOCTRL;

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
    self.BleCentralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 1 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        [self scanDevice];
    });
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table View

- (IBAction)segmentAction:(id)sender
{
    UITableViewCell* cell = [[sender superview] superview];
    NSIndexPath *indexPath =
    [self.TableView indexPathForCell:cell];
    NSUInteger row = indexPath.row;
    IOCTRL data = {row, 2};
    if (self.ConnPeri && self.ConnChar)
        [self.ConnPeri writeValue:[NSData dataWithBytes:&data length:2] forCharacteristic:self.ConnChar
                                   type:CBCharacteristicWriteWithoutResponse];
}

- (IBAction)switchAction:(id)sender
{
    UITableViewCell* cell = [[sender superview] superview] ;
    NSIndexPath *indexPath =
    [self.TableView indexPathForCell:cell];
    NSUInteger row = indexPath.row;
    NSInteger  onoff = 0;
    if (((UISwitch*)sender).on)
        onoff = 1;
    IOCTRL data = { row, onoff};
    if (self.ConnPeri && self.ConnChar)
        [self.ConnPeri writeValue:[NSData dataWithBytes:&data length:2] forCharacteristic:self.ConnChar
                         type:CBCharacteristicWriteWithoutResponse];
    
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;//[[self.fetchedResultsController sections] count];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    //    id <NSFetchedResultsSectionInfo> sectionInfo = [self.fetchedResultsController sections][section];
    return 32;//[sectionInfo numberOfObjects];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    UILabel *label = (UILabel *)[cell viewWithTag:1];
    label.text = [NSString stringWithFormat:@"%2ld : ", (long)indexPath.row];
    UISegmentedControl *segCtrl = (UISegmentedControl*)[cell viewWithTag:2];
    [segCtrl addTarget:self action:@selector(segmentAction:) forControlEvents:UIControlEventValueChanged];
    UISwitch *switchCtrl = (UISwitch*)[cell viewWithTag:3];
    [switchCtrl addTarget:self action:@selector(switchAction:) forControlEvents:UIControlEventValueChanged];
    //    [self configureCell:cell atIndexPath:indexPath];
    /*EkoCCBlue *object = _objects[indexPath.row];
    if (object)
    {
        NSLog(@"Device name %@", object.Name);
        cell.textLabel.text = object.Name;//[NSString stringWithFormat:@"%@", [object Name]];
    }*/
    
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return NO;
}


//
//  Bluetooth stuffs start here
//

- (void) scanDevice
{
//    [_objects removeAllObjects];
//    [self.tableView reloadData];
    
    if (_BleCentralManager.state == CBCentralManagerStatePoweredOn)
    {
        
        // Scans for any peripheral
        CBUUID *serviceUUID = [CBUUID UUIDWithString:BLINKY_SERVICE_UUID];
        
        NSArray *peripherals = [_BleCentralManager retrieveConnectedPeripheralsWithServices:@[serviceUUID]];
        if (peripherals.count <= 0)
            [self.BleCentralManager scanForPeripheralsWithServices:@[serviceUUID] options:nil];
        else
        {
            NSLog(@"Found Connected peripheral %@", peripherals);
            //            [_DeviceList addObject:peripherals[0]];//[NSString stringWithFormat:@"%@", peripheral.name]];
            //            [_DevicePickerView reloadAllComponents];
            for (CBPeripheral *peripheral in peripherals)
            {
//                _CurrPeripheral = peripheral;
                //peripheral.delegate = self;
                // [self.BleCentralManager connectPeripheral:peripheral options:nil];
            }
        }
    }
}

- (void)centralManager:(CBCentralManager *)central didDiscoverPeripheral:(CBPeripheral *)peripheral advertisementData:(NSDictionary *)advertisementData RSSI:(NSNumber *)RSSI {

    // Stops scanning for peripheral
    [self.BleCentralManager stopScan];
    
    peripheral.delegate = self;
    self.ConnPeri = peripheral;
    
    // Connects to the discovered peripheral
    [self.BleCentralManager connectPeripheral:peripheral options:nil];
}

- (void)centralManager:(CBCentralManager *)central didConnectPeripheral:(CBPeripheral *)peripheral
{
    CBUUID *servUUID = [CBUUID UUIDWithString:BLINKY_SERVICE_UUID];
    
    [peripheral discoverServices:@[servUUID]];
    NSLog(@"Peripheral connected");
}

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverServices:(NSError *)error
{
    CBUUID *serviceUUID = [CBUUID UUIDWithString:BLINKY_SERVICE_UUID];
    CBUUID *charUUID = [CBUUID UUIDWithString:BLINKY_CHAR_UUID];
    
    for (CBService *service in peripheral.services) {
        NSLog(@"Discovered service %@", service);
        
        if ([service.UUID isEqual:(serviceUUID)])
            [peripheral discoverCharacteristics:@[charUUID] forService:service];
    }
}

- (void)peripheral:(CBPeripheral *)peripheral didDiscoverCharacteristicsForService:(CBService *)service error:(NSError *)error
{
    if (error) {
        NSLog(@"Error discovering characteristics: %@", [error localizedDescription]);
        //        [self cleanup];
        return;
    }
    
    CBUUID *charUUID = [CBUUID UUIDWithString:BLINKY_CHAR_UUID];
    for (CBCharacteristic *characteristic in service.characteristics)
    {
        NSLog(@"Discovered characteristic %@", characteristic);
        if ([characteristic.UUID isEqual:(charUUID)])
        {
            self.ConnChar = characteristic;
            if (characteristic.properties & CBCharacteristicPropertyRead)
                [peripheral readValueForCharacteristic:characteristic];
        }
    }
    
}

- (void)peripheral:(CBPeripheral *)peripheral didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic error:(NSError *)error
{
    if (!error)
    {
        
        CBUUID *charUUID = [CBUUID UUIDWithString:BLINKY_CHAR_UUID];//DEVINFO_SYSID_UUID];
        
        if ([characteristic.UUID isEqual:(charUUID)])
        {
            int64_t d = *(int64_t*)([characteristic.value bytes]);
            //NSString* name = [NSString stringWithFormat:@"%@-%llX", peripheral.name, d];
            //EkoCCBlue *dev = [[EkoCCBlue alloc] init:name peripheral:peripheral];
            //if (!_objects) {
            //    _objects = [[NSMutableArray alloc] init];
            //}
            //[_objects insertObject:dev atIndex:0];
            
            //_CurrPeripheral = nil;
            
            //[self.tableView reloadData];
            
            //NSIndexPath *indexPath = [NSIndexPath indexPathForRow:0 inSection:0];
            //[self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
            //[self.BleCentralManager cancelPeripheralConnection:peripheral];
        }
    }
}

- (void)centralManager:(CBCentralManager *)central didDisconnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error
{
    
    //[self.detailViewController deviceDisconnected];
    [self scanDevice];
}

- (void)centralManager:(CBCentralManager *)central didRetrieveConnectedPeripherals:(NSArray *)peripherals
{
    NSLog(@"Connected peripheral %@", peripherals);
    
}

- (void)centralManagerDidUpdateState:(CBCentralManager *)central
{
    
}



@end

File: SensorTagDemo/SensorTagDemo/ViewController.swift
//
//  ViewController.swift
//  SensorTagDemo
//
//  Created by Nguyen Hoan Hoang on 2017-10-08.
//  Copyright © 2017 I-SYST inc. All rights reserved.
//

import UIKit
import CoreBluetooth
import simd

class ViewController: UIViewController, CBCentralManagerDelegate {

    var bleCentral : CBCentralManager!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        bleCentral = CBCentralManager(delegate: self, queue: DispatchQueue.main)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
        
    }

    @IBOutlet weak var graphView : GraphView!
    @IBOutlet weak var tempLabel: UILabel!
    @IBOutlet weak var humiLabel : UILabel!
    @IBOutlet weak var pressLabel : UILabel!
    @IBOutlet weak var rssiLabel : UILabel!
    @IBOutlet weak var aqiLabel : UILabel!
    @IBOutlet weak var batLevelLabel : UILabel!
    @IBOutlet weak var batVoltLabel : UILabel!
    
    // MARK: BLE Central
    
    func centralManager(_ central: CBCentralManager,
                        didDiscover peripheral: CBPeripheral,
                        advertisementData : [String : Any],
                        rssi RSSI: NSNumber) {
        print("PERIPHERAL NAME: \(String(describing: peripheral.name))\n AdvertisementData: \(advertisementData)\n RSSI: \(RSSI)\n")
        
        print("UUID DESCRIPTION: \(peripheral.identifier.uuidString)\n")
        
        print("IDENTIFIER: \(peripheral.identifier)\n")
        
        if advertisementData[CBAdvertisementDataManufacturerDataKey] == nil {
            return
        }
        
        //sensorData.text = sensorData.text + "FOUND PERIPHERALS: \(peripheral) AdvertisementData: \(advertisementData) RSSI: \(RSSI)\n"
        var manId = UInt16(0)
        let manData = advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData
        
        if manData.length < 3 {
            return
        }
        //(advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&manId, range: NSMakeRange(0, 2))
        manData.getBytes(&manId, range: NSMakeRange(0, 2))
        if manId != 0x177 {
            return
        }
        
        var type = UInt8(0)
        //(advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&type, range: NSMakeRange(2, 1))
        manData.getBytes(&type, range: NSMakeRange(2, 1))
        
        switch (type) {
        	case 1:
                var press = Int32(0)
                (advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&press, range: NSMakeRange(3, 4))
                pressLabel.text = String(format:"%.2f KPa", Float(press) / 1000.0)
                
                var temp = Int16(0)
                (advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&temp, range: NSMakeRange(7, 2))
                tempLabel.text = String(format:"%.2f C", Float(temp) / 100.0)
                
                var humi = UInt16(0)
                (advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&humi, range: NSMakeRange(9, 2))
                humiLabel.text = String(format:"%d%%", humi / 100)
                graphView.add(double3(Double(temp) / 100.0, Double(press) / 1000.0, Double(humi) / 100.0))
            	break
        	case 2:
                var aqi = UInt16(0)
                (advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&aqi, range: NSMakeRange(3, 2))
            	break
        	case 16:
                var batLevel = UInt8(0)
                (advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&batLevel, range: NSMakeRange(3, 1))
                batLevelLabel.text = String(format:"%d%%", batLevel)

                var batVolt = Int32(0)
                (advertisementData[CBAdvertisementDataManufacturerDataKey] as! NSData).getBytes(&batVolt, range: NSMakeRange(4, 4))
                batVoltLabel.text = String(format:"%.2fV", Float(batVolt) / 1000.0)
                
            	break
        	default:
            	break
        }

        rssiLabel.text = String( describing: RSSI)
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        
        peripheral.discoverServices(nil)
        print("Connected to peripheral")
        
        
    }
    
    func centralManager(_ central: CBCentralManager,
                        didDisconnectPeripheral peripheral: CBPeripheral,
                        error: Error?) {
        print("disconnected from peripheral")
        
        
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
    }
    
    func scanPeripheral(_ sender: CBCentralManager)
    {
        print("Scan for peripherals")
        bleCentral.scanForPeripherals(withServices: nil, options: nil)
    }

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
            
        case .poweredOff:
            print("CoreBluetooth BLE hardware is powered off")
            break
        case .poweredOn:
            print("CoreBluetooth BLE hardware is powered on and ready")
            
            bleCentral.scanForPeripherals(withServices: nil, options: nil)
            break
        case .resetting:
            print("CoreBluetooth BLE hardware is resetting")
            break
        case .unauthorized:
            print("CoreBluetooth BLE state is unauthorized")
            
            break
        case .unknown:
            print("CoreBluetooth BLE state is unknown")
            break
        case .unsupported:
            print("CoreBluetooth BLE hardware is unsupported on this platform")
            break
            
        }
    }
    
}

